docs/keras.rst

```diff
@@ -1,7 +1,130 @@
 Horovod with Keras
 ==================
-Horovod supports Keras and regular TensorFlow in similar ways.
+Horovod supports Keras and regular TensorFlow in similar ways. To use Horovod with Keras, make the following additions to your program. 
 
-See full training `simple <https://github.com/horovod/horovod/blob/master/examples/keras_mnist.py>`_ and `advanced <https://github.com/horovod/horovod/blob/master/examples/keras_mnist_advanced.py>`_ examples.
+1. Run ``hvd.init()``.
 
-.. NOTE:: Keras 2.0.9 has a `known issue <https://github.com/fchollet/keras/issues/8353>`_ that makes each worker allocate all GPUs on the server, instead of the GPU assigned by the *local rank*. If you have multiple GPUs per server, upgrade to Keras 2.1.2 or downgrade to Keras 2.0.8.
\ No newline at end of file
+2. Pin a server GPU to be used by this process using ``config.gpu_options.visible_device_list``.
+
+   With the typical setup of one GPU per process, you can set this to *local rank*. In that case, the first process on
+   the server will be allocated the first GPU, the second process will be allocated the second GPU, and so forth.
+
+3. Scale the learning rate by the number of workers.
+
+   Effective batch size in synchronous distributed training is scaled by the number of workers.
+   An increase in learning rate compensates for the increased batch size.
+
+4. Wrap the optimizer in ``hvd.DistributedOptimizer``.
+
+   The distributed optimizer delegates gradient computation to the original optimizer, averages gradients using *allreduce* or *allgather*, and then applies those averaged gradients.
+
+5. Add ``hvd.callbacks.BroadcastGlobalVariablesCallback(0)`` to broadcast initial variable states from rank 0 to all other processes.
+
+   This is necessary to ensure consistent initialization of all workers when training is started with random weights or restored from a checkpoint.
+
+6. Modify your code to save checkpoints only on worker 0 to prevent other workers from corrupting them.
+
+   Accomplish this by guarding model checkpointing code with ``hvd.rank() != 0``.
+
+.. NOTE:: Keras 2.0.9 has a `known issue <https://github.com/fchollet/keras/issues/8353>`_ that makes each worker allocate all GPUs on the server, instead of the GPU assigned by the *local rank*. If you have multiple GPUs per server, upgrade to Keras 2.1.2 or downgrade to Keras 2.0.8.
+
+See full training `simple <https://github.com/horovod/horovod/blob/master/examples/keras_mnist.py>`_ (shown below) and `advanced <https://github.com/horovod/horovod/blob/master/examples/keras_mnist_advanced.py>`_ examples.
+
+
+.. code-block:: python
+
+    from __future__ import print_function
+    import keras
+    from keras.datasets import mnist
+    from keras.models import Sequential
+    from keras.layers import Dense, Dropout, Flatten
+    from keras.layers import Conv2D, MaxPooling2D
+    from keras import backend as K
+    import math
+    import tensorflow as tf
+    import horovod.keras as hvd
```

Comment:
Maybe let's also document that to use TF-Keras, user should do `from tensorflow import keras` instead of `import keras` and `import horovod.tensorflow.keras as hvd` instead of `import horovod.keras as hvd`?

Or should we make a separate section for this?

Reply:
Good idea, since this would make the imports more clear. I don't think it is necessary to make a separate section for something so simple, so if you agree I'll change this example. :)

Reply:
What are you thinking w.r.t. change, add TF-Keras as an alternative commented out?  If so, I think we should also add **NOTE** before/after to draw additional attention to those changes to be made if switching to TF-Keras.